firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -126,101 +127,133 @@ String getName() {
+    }
 
-    if (cacheEntryValue == null
-        || cacheEntryValue.getCacheStatus() == FiidCache.CacheStatus.REGISTER_ERROR) {
-      String fid = Utils.createRandomFid();
+    return null;
+  }
 
-      boolean firstUpdateCacheResult =
-          localCache.insertOrUpdateCacheEntry(
-              FiidCacheEntryValue.create(fid, FiidCache.CacheStatus.UNREGISTERED, "", "", 0, 0));
+  private static boolean persistedFidDoesNotExistsOrInErrorState(
+      PersistedFidEntryValue persistedFidEntryValue) {
+    return persistedFidEntryValue == null
+        || persistedFidEntryValue.getPersistedStatus()
+            == PersistedFid.PersistedStatus.REGISTER_ERROR;
+  }
 
-      if (!firstUpdateCacheResult) {
-        throw new FirebaseInstallationsException(
-            "Failed to update client side cache.",
-            FirebaseInstallationsException.Status.CLIENT_ERROR);
-      }
+  private static boolean persistedFidExistsAndRegistered(
+      PersistedFidEntryValue persistedFidEntryValue) {
+    return persistedFidEntryValue != null
+        && persistedFidEntryValue.getPersistedStatus() == PersistedFid.PersistedStatus.REGISTERED;
+  }
 
-      registerAndSaveFID(fid);
+  private static boolean persistedFidExistsAndUnregistered(
+      PersistedFidEntryValue persistedFidEntryValue) {
+    return persistedFidEntryValue != null
+        && persistedFidEntryValue.getPersistedStatus() == PersistedFid.PersistedStatus.UNREGISTERED;
+  }
 
-      return fid;
-    } else if (!cacheEntryValue.getFirebaseInstallationId().isEmpty()) {
+  private String createFidAndPersistFidEntry() throws FirebaseInstallationsException {
+    String fid = Utils.createRandomFid();
 
-      if (cacheEntryValue.getCacheStatus() == FiidCache.CacheStatus.UNREGISTERED) {
-        registerAndSaveFID(cacheEntryValue.getFirebaseInstallationId());
-      }
+    boolean firstUpdateCacheResult =
+        persistedFid.insertOrUpdatePersistedFidEntry(
+            PersistedFidEntryValue.builder()
+                .setFirebaseInstallationId(fid)
+                .setPersistedStatus(PersistedFid.PersistedStatus.UNREGISTERED)
+                .setTokenCreationEpochInSecs(0)
+                .setExpiresInSecs(0)
+                .build());
 
-      return cacheEntryValue.getFirebaseInstallationId();
+    if (!firstUpdateCacheResult) {
+      throw new FirebaseInstallationsException(
+          "Failed to update client side cache.",
+          FirebaseInstallationsException.Status.CLIENT_ERROR);
     }
-    return null;
+
+    registerAndSaveFid(fid);
+
+    return fid;
   }
 
   /**
-   * Registers the created FID with FIS Servers if the Network is available and update the cache.
+   * Registers the created Fid with FIS Servers if the Network is available and update the shared
+   * prefs.
    */
-  private void registerAndSaveFID(String fid) throws FirebaseInstallationsException {
+  private void registerAndSaveFid(String fid) throws FirebaseInstallationsException {
```

Comment:
This method has to be asynchronous to not block callers to #getId. 
What I'm saying is that if someone calls #getId, the method should immediately return an FID independent of if that FID has to be registered with FIS or not.

Reply:
Done. Please take another look.